advanced-game-design-team/Artifacts/Raycast Collisions/README.md

```diff
@@ -0,0 +1,112 @@
+# Raycasted Point Collision
+
+![](./rays.gif)
+
+## Introduction
+
+This guide explores a 2D raycasting system designed for detecting collisions between rays emanating from a point (typically the mouse cursor) and rectangular obstacles. The system comprises several key components that work together to create an efficient collision detection mechanism.
+
+## Core Components
+
+### 1. Ray Generation
+
+The system generates rays from the mouse cursor position in a 360-degree circle. The number of rays is defined by `NUM_RAYS`, which is set to 360 in this implementation, effectively creating one ray per degree.
+
+```cpp
+struct ray {
+    point_2d origin;
+    vector_2d direction;
+};
+
+// In the main loop:
+point_2d mouse = mouse_position();
+std::vector<ray_hit> hits = cast_rays(obstacles, mouse, NUM_RAYS, MAX_DISTANCE);
+```
+
+### 2. Obstacle Generation
+
+Rectangular obstacles are randomly generated at the start of the program, creating a dynamic environment for ray collision:
+
+```cpp
+std::vector<rectangle> generate_random_obstacles(int num_obstacles) {
+    std::vector<rectangle> obstacles;
+    for (int i = 0; i < num_obstacles; ++i) {
+        double x = rnd(SCREEN_WIDTH);
+        double y = rnd(SCREEN_HEIGHT);
+        double width = rnd(10, 300);
+        double height = rnd(10, 20);
+        obstacles.push_back(rectangle_from(x, y, width, height));
+    }
+    return obstacles;
+}
+```
+
+### 3. Ray-Rectangle Intersection Detection
+
+The core of the collision detection is the `ray_rectangle_intersection` function, which implements the "slab" method for Ray-AABB (Axis-Aligned Bounding Box) intersection:
+
+```cpp
+bool ray_rectangle_intersection(const ray& r, const rectangle& rect, ray_hit& hit) {
+    // Implementation details...
+}
+```
+
+This function calculates the intersection point (if any) between a ray and a rectangle, storing the result in the `ray_hit` structure.
+
+### 4. Ray Marching
+
+The `march_ray` function simulates ray marching by checking for intersections with all obstacles and returning the closest hit  (this is not a real march, we don't iteratively extend a vector we just check for intersections and take the closest one):
+
+```cpp
+ray_hit march_ray(const ray& r, const std::vector<rectangle>& obstacles, double max_distance) {
+    // Implementation details...
+}
+```
+
+### 5. Visualization
+
+The system visualizes the rays and obstacles using SplashKit's drawing functions:
```

Comment:
I'd like to a more explicit explanation of how this goes from abstract code to the visual result. I think it needs to be put together more explicitly.

E.g. something with this vibe:
"Now that we've already calculated the collision point for all of our rays, we draw lines from the origin point to each of these collision points. Rays with no collision point continue to the edge of the render area since nothing is blocking them.

Lastly we draw the rectangle obstacles which visually represent the objects causing the ray collisions."